src/ngraph_assign_clusters.cc

```diff
@@ -559,12 +653,88 @@ Status AssignClusters(Graph* graph) {
 
       // TODO(amprocte): move attr name to a constant
       node->AddAttr("_ngraph_cluster", cluster_idx);
+
+      if (config::IsLoggingPlacement()) {
+        // map from cluster id to ngraph_cluster id
+        cluster_to_encapsulate[cluster->index] = cluster_idx;
+      }
     }
 
     seen.insert(cluster);
   }
   NGRAPH_VLOG(2) << "Tagging done";
 
+  if (config::IsLoggingPlacement()) {
+    std::cout << "\n=============Edge contraction logs=============\n";
+    int num_reasons = 7;  // the number of elements in the reasons enum
+    vector<int> reason_count(num_reasons,
+                             0);  // histogram of reasons of non-contraction
+    int num_non_contracted = 0;
+    std::vector<string> reason_string(  // to convert the enum to string
+        {"NOTANOP", "UNSUPPORTED", "DEADNESS", "BACKEND", "SAMECLUSTER",
+         "STATICINPUT", "PATHEXISTS"});
+    std::cout
+        << "_ngraph_cluster i->j: non contraction reason histogram (Cannot be "
+           "UNSUPPORTED or NOTANOP because unsupported ops will not be "
+           "assigned an "
+           "encapsulate)\n";
+    for (auto it : cluster_separation_reason) {
+      num_non_contracted += it.second.size();
+      vector<string> reasons_for_this_edge;
+      for (auto& inner_itr : it.second) {
+        reason_count[inner_itr]++;
+        reasons_for_this_edge.push_back(reason_string[inner_itr]);
+      }
+      auto cluster_id_vector = ng::split(it.first, ',');
+      // function to find if this cluster became an ngraph_cluster
+      // returns ngraph_cluster id if yes, else returns -1
+      auto find_in_map = [&cluster_to_encapsulate, &cluster_id_vector](int x) {
+        auto itr = cluster_to_encapsulate.find(stoi(cluster_id_vector[x]));
+        return itr == cluster_to_encapsulate.end() ? -1 : itr->second;
+      };
+      int src_encapsulate = find_in_map(0);
+      int dst_encapsulate = find_in_map(1);
+      bool both_src_dst_are_encapsulates =
+          src_encapsulate >= 0 && dst_encapsulate >= 0;
+      bool src_dst_are_distinct = src_encapsulate != dst_encapsulate;
+      if (both_src_dst_are_encapsulates && src_dst_are_distinct) {
```

Comment:
If both are encapsulated they will always be distinct? Do we need the src_dst_are_distinct check?

Reply:
Yes, we always expect them to be distinct. But if they are ever not, then we will miss counting one edge which will later get caught in the "count total edges vs num_edges()" assert, so we will know.

Reply:
The endpoints of an edge could lie in the same cluster. Without this I see prints that say "2,2 did not merge because of PATHEXIXTS" etc